moveit_experimental/jog_arm/src/jog_arm/jog_ros_interface.cpp

```diff
@@ -0,0 +1,1243 @@
+        (ini_condition < parameters_.hard_stop_singularity_threshold))
+    {
+      velocity_scale = 1. -
+                       (ini_condition - parameters_.lower_singularity_threshold) /
+                           (parameters_.hard_stop_singularity_threshold - parameters_.lower_singularity_threshold);
+    }
+
+    // Very close to singularity, so halt.
+    else if (ini_condition > parameters_.hard_stop_singularity_threshold)
+    {
+      velocity_scale = 0;
+      ROS_WARN_NAMED(LOGNAME, "Close to a singularity. Halting.");
+    }
+  }
+
+  return velocity_scale;
+}
+
+bool JogCalcs::checkIfJointsWithinBounds(trajectory_msgs::JointTrajectory& new_jt_traj)
+{
+  bool halting = false;
+  for (auto joint : joint_model_group_->getJointModels())
+  {
+    if (!kinematic_state_->satisfiesVelocityBounds(joint))
+    {
+      ROS_WARN_STREAM_THROTTLE_NAMED(2, LOGNAME, ros::this_node::getName() << " " << joint->getName() << " "
+                                                                           << " close to a "
+                                                                              " velocity limit. Enforcing limit.");
+      kinematic_state_->enforceVelocityBounds(joint);
+      for (std::size_t c = 0; c < new_jt_traj.joint_names.size(); ++c)
+      {
+        if (new_jt_traj.joint_names[c] == joint->getName())
+        {
+          new_jt_traj.points[0].velocities[c] = kinematic_state_->getJointVelocities(joint)[0];
+          break;
+        }
+      }
+    }
+
+    // Halt if we're past a joint margin and joint velocity is moving even
+    // farther past
+    double joint_angle = 0;
+    for (std::size_t c = 0; c < new_jt_traj.joint_names.size(); ++c)
+    {
+      if (original_jt_state_.name[c] == joint->getName())
+      {
+        joint_angle = original_jt_state_.position.at(c);
+        break;
+      }
+    }
+
+    if (!kinematic_state_->satisfiesPositionBounds(joint, -parameters_.joint_limit_margin))
+    {
+      const std::vector<moveit_msgs::JointLimits> limits = joint->getVariableBoundsMsg();
+
+      // Joint limits are not defined for some joints. Skip them.
+      if (limits.size() > 0)
+      {
+        if ((kinematic_state_->getJointVelocities(joint)[0] < 0 &&
+             (joint_angle < (limits[0].min_position + parameters_.joint_limit_margin))) ||
+            (kinematic_state_->getJointVelocities(joint)[0] > 0 &&
+             (joint_angle > (limits[0].max_position - parameters_.joint_limit_margin))))
+        {
+          ROS_WARN_STREAM_THROTTLE_NAMED(2, LOGNAME, ros::this_node::getName() << " " << joint->getName()
+                                                                               << " close to a "
+                                                                                  " position limit. Halting.");
+          halting = true;
+        }
+      }
+    }
+  }
+
+  return !halting;
+}
+
+void JogCalcs::publishWarning(const bool active) const
```

Comment:
bools (all PODs) do not need const when passed in

Reply:
This confuses me: why don't plain-old-data types need `const`? Do you mean primitive types? That would still confuse me as it would not be universally true.

Reply:
a POD variable is copied when passed in, so the ``const`` doesn't do anything and by common convention is left out. 

C++03: 
> Section 3.9.1 lists bool as an integral type, and although not stated explicitly, it is reasonable to deduce that integral types are arithmetic types. Consequently, bool is a POD type.

Reply:
@AndyZe overall this really doesn't matter, I'm just giving you a Google-style code review and maybe being too picky